### PR TITLE
feat: Added dashboard for Alert Quality

### LINKED ISSUE #1779: [➕ Feature]: Alert quality widgets for a dashboard
Users connect a lot of providers for Keep, and we should provide some analytics about the quality of events coming out of those providers.

I propose to add a few widgets, and an auto-provisioned dashboard called "Alert quality". 

Imagine a use-case when users wants to make sure that all alert providers provide "severity", "department", "product" and "affected_customer" from the sources. This dashboard should help them.

Widgets:
- A table with a per-provider alert quality. Rows: connected providers. Columns: provider name, amount of alerts received, % of alerts correlated to incidents, % of alerts having a "field" filled. Where field is a configurable set of fields. 
- % of alerts correlated to incidents
- % of alerts having a "field" filled

### PR DESCRIPTION
**Pull request by vikashsprem**
Closes #1779
/claim #1779 
## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

**Comment by Matvey-Kuk** [comment]
Hey @vikashsprem how is it going? :)

**Comment by vikashsprem** [comment]
Hey, @Matvey-Kuk going great, need some time to complete it.

**Comment by vikashsprem** [comment]
Hey @Matvey-Kuk, Please review this PR and let me know if any changes.

Screenshot 2024-09-27 at 1 15 06 PM

**Comment by vikashsprem** [comment]
Thanks for the suggestion! My concern with making it a widget is that the separate page allows us to include more detailed functionality without crowding the main dashboard. But I'm happy to discuss and find the best approach!

**Comment by Matvey-Kuk** [comment]
@vikashsprem the main dashboard has a timeframe selector, also we're planning to add more widgets to this dashboard in the future to allow users to tail dashboards to their needs.

**Comment by vikashsprem** [comment]
@Matvey-Kuk  Got it, that makes sense! I'll look into transitioning the separate page into a widget.

**Comment by talboren** [comment]
Hey @vikashsprem 

Please note that the build fails
`158:17  Error: React Hook "useEffect" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function.  react-hooks/rules-of-hooks `

Check it out with `npm run build`

**Comment by vikashsprem** [comment]
> Hey @vikashsprem
> 
> Please note that the build fails `158:17 Error: React Hook "useEffect" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function. react-hooks/rules-of-hooks `
> 
> Check it out with `npm run build`

Okay let me check, I going to fix it now.

**Comment by vikashsprem** [comment]
Hey @talboren, Now you can check it out it has been fixed now.

**Comment by talboren** [comment]
@vikashsprem please note that now you have some linters failing
CleanShot 2024-09-29 at 09 40 19

**Comment by rajeshj11** [comment]
steps:
1. Go to dashboards.
2. create a dashboard or use an existing dashboard.
3. click on the add widget.
4. choose widget type as generic metrics
5. select the generic metrics(right now we have one generic report)
6. and click on add widget.

expected:
should able to see the new alert quality report in the dashbaord

**Comment by vikashsprem** [comment]
@Matvey-Kuk @talboren As you suggested we have implemented. Now it's ready for review.

https://github.com/user-attachments/assets/0e44415c-2f4b-4c2d-a36b-5744cd12ad7b

**Comment by vikashsprem** [comment]
Hey @talboren @Matvey-Kuk  is the test (Tests / tests (pull_request) ) failing from our end?

**Review by talboren — CHANGES_REQUESTED** [review]
Please see my last comment

**Comment by rajeshj11** [comment]
1. Custom Metric doesn't work, what is it?
    it is just another widget type. not implemented(we can remove if we want to)
2. Fields are strange, it doesn't show all the fields like in the feed
    Regarding the fields. we have restricted only three fields. @Matvey-Kuk also agreed with it. 

3. The time range is weird and user cannot understand what time frame is he looking at. it should be "everything" unless specified otherwise (unless timeframe is provided I mean)
 Currently, if time frame is not mentioned we are taking lat seven days. this we need to rework. will work on it
 
4. The table shows providers by the provider type. In case I have 5 Grafana's it shows Grafana 5 times. It should group by the provider id and also mention that in the table so I can identify which provider exactly.

I am confused here. we are grouping by product_type. in the above screen share how come 5 grafana have different product_type? If we get the SQL data that would more useful to replicate the issue. (we will also again check the aggregation query and get back to you)

**Comment by rajeshj11** [comment]
> Hey @vikashsprem I'm adding my comments here:
> 
> 1. Custom Metric doesn't work, what is it?
> 2. Fields are strange, it doesn't show all the fields like in the feed
> 3. The time range is weird and user cannot understand what time frame is he looking at. it should be "everything" unless specified otherwise (unless timeframe is provided I mean)
> 4. The table shows providers by the provider type. In case I have 5 Grafana's it shows Grafana 5 times. It should group by the provider id and also mention that in the table so I can identify which provider exactly.
> 
>  CleanShot.2024-10-07.at.11.41.43.mp4 
>  CleanShot.2024-10-07.at.11.40.24.mp4


1. Custom Metric doesn't work, what is it?
    it is just another widget type. not implemented(we can remove if we want to)
2. Fields are strange, it doesn't show all the fields like in the feed
    Regarding the fields. we have restricted only three fields. @Matvey-Kuk also agreed with it. 

3. The time range is weird and user cannot understand what time frame is he looking at. it should be "everything" unless specified otherwise (unless timeframe is provided I mean)
 Currently, if time frame is not mentioned we are taking lat seven days. this we need to rework. will work on it
 
4. The table shows providers by the provider type. In case I have 5 Grafana's it shows Grafana 5 times. It should group by the provider id and also mention that in the table so I can identify which provider exactly.
   I am confused here. we are grouping by product_type. in the above screen share how come 5 grafana have different 
   product_type? If we get the SQL data that would more useful to replicate the issue. (we will also again check the aggregation 
   query and get back to you)

**Comment by talboren** [comment]
> > Hey @vikashsprem I'm adding my comments here:
> > 
> > 1. Custom Metric doesn't work, what is it?
> > 2. Fields are strange, it doesn't show all the fields like in the feed
> > 3. The time range is weird and user cannot understand what time frame is he looking at. it should be "everything" unless specified otherwise (unless timeframe is provided I mean)
> > 4. The table shows providers by the provider type. In case I have 5 Grafana's it shows Grafana 5 times. It should group by the provider id and also mention that in the table so I can identify which provider exactly.
> > 
> > CleanShot.2024-10-07.at.11.41.43.mp4
> > CleanShot.2024-10-07.at.11.40.24.mp4
> 
> 1. Custom Metric doesn't work, what is it?
>    it is just another widget type. not implemented(we can remove if we want to)
> 2. Fields are strange, it doesn't show all the fields like in the feed
>    Regarding the fields. we have restricted only three fields. @Matvey-Kuk also agreed with it.
> 3. The time range is weird and user cannot understand what time frame is he looking at. it should be "everything" unless specified otherwise (unless timeframe is provided I mean)
>    Currently, if time frame is not mentioned we are taking lat seven days. this we need to rework. will work on it
> 4. The table shows providers by the provider type. In case I have 5 Grafana's it shows Grafana 5 times. It should group by the provider id and also mention that in the table so I can identify which provider exactly.
>    I am confused here. we are grouping by product_type. in the above screen share how come 5 grafana have different
>    product_type? If we get the SQL data that would more useful to replicate the issue. (we will also again check the aggregation
>    query and get back to you)

Re #1: Please remove it if it's not implemented.
Re #2: If @Matvey-Kuk confirmed this then OK.
Re #3: It should *at least* mention that it's last 7 days. Changing the timeframe does nothing at the moment (you have no way to know if it changed something or not)
Re #4: I honestly didn't understand. You can connect more than 1 tool of the same type to Keep. So for instance, I can have Grafana A sending alerts, Grafana B sending alerts, Grafana C sending alerts. Now we show the same data for the *provider type* therefore the data in the table is false.

Let me know if I can help anyhow!

cc: @Matvey-Kuk @vikashsprem @rajeshj11

**Comment by vikashsprem** [comment]
@talboren  As you suggested @rajeshj11 has made some changes. Could you please review the PR?

Screenshot 2024-10-07 at 7 02 06 PM

**Comment by vikashsprem** [comment]
Hey @talboren @Matvey-Kuk It's ready for review. Please let us know if you have any concerns.

**Comment by rajeshj11** [comment]
@talboren @Matvey-Kuk We have worked on the feedback changes. can you please review now

**Comment by Matvey-Kuk** [comment]
Will review shortly!

File: keep-ui/app/alerts/quality/page.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import QualityAlert from './alert-quality-table'; // Adjust the import path if needed
+
+const AlertQualityPage = () => {
```
Comment (Matvey-Kuk) [inline keep-ui/app/alerts/quality/page.tsx:3]: I guess we don't need this page anymore, right?

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]
For some reason it doesn't render for me even though the API call seems to be working perfecly:
Screenshot 2024-10-13 at 10 43 09

File: keep-ui/app/alerts/quality/page.tsx (hunk 1 above)
Comment (rajeshj11) [reply keep-ui/app/alerts/quality/page.tsx:3]: Page is not needed. but component is needed

**Comment by rajeshj11** [comment]
> For some reason it doesn't render for me even though the API call seems to be working perfecly: Screenshot 2024-10-13 at 10 43 09
> 
> To reproduce:
> 
> 1. Remove db.
> 2. Shoot scripts/simulate_alerts.py
> 3. Create dashboard, add the widget.

Will check

**Comment by rajeshj11** [comment]
> > For some reason it doesn't render for me even though the API call seems to be working perfecly: Screenshot 2024-10-13 at 10 43 09
> > To reproduce:
> > 
> > 1. Remove db.
> > 2. Shoot scripts/simulate_alerts.py
> > 3. Create dashboard, add the widget.
> 
> Will check

Mostly post deletion. You have not installed any providers. We are showing only installed metrics
Cc: @talboren

File: keep-ui/app/alerts/quality/page.tsx (hunk 1 above)
Comment (rajeshj11) [reply keep-ui/app/alerts/quality/page.tsx]: resolved

**Comment by rajeshj11** [comment]
@Matvey-Kuk @talboren  I wanted to follow up on this PR, as it seems to have been pending for a while.

**Comment by Matvey-Kuk** [comment]
@rajeshj11 sorry for late reply! I've checked with Tal, the thing is that once the alert pops up in Keep and it doesn't have an Installed provider, a new type of providers is getting created called "Linked Providers", they are acting as any other providers, and I think they should be represented in the table.

Could you please update the PR?

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]
@rajeshj11 sorry for late reply! I've checked with Tal, the thing is that once the alert pops up in Keep and it doesn't have an Installed provider, a new type of providers is getting created called "Linked Providers", they are acting as any other providers, and I think they should be represented in the table.

Could you please update the PR?

**Comment by rajeshj11** [comment]
> @rajeshj11 sorry for late reply! I've checked with Tal, the thing is that once the alert pops up in Keep and it doesn't have an Installed provider, a new type of providers is getting created called "Linked Providers", they are acting as any other providers, and I think they should be represented in the table.
> 
> Could you please update the PR?

@Matvey-Kuk I did discuss this with @talboren. I proposed I will show all the provider type stats in separate table. Something like below.

Github(linked +installed) in a separate table.
Gitlab(linked+installed)

Should I maintain separate table. Need conformation

**Comment by Matvey-Kuk** [comment]
@rajeshj11 I didn't mean a separate table, but just to include alerts from Linked + Installed in the widget you already added :)

**Comment by Matvey-Kuk** [comment]
@rajeshj11 looks cool!

**Comment by vikashsprem** [comment]
@Matvey-Kuk the latest change has been applied, Please review.

**Review by Matvey-Kuk — CHANGES_REQUESTED** [review]
Great! I noticed a bug:
1) I click "Add Fields", add a few fields.
2) I save dashboard.
3) I reload the page.
4) I don't see fields I added.

Screenshot 2024-10-20 at 16 06 58

**Comment by rajeshj11** [comment]
> Great! I noticed a bug:
> 
> 1. I click "Add Fields", add a few fields.
> 2. I save dashboard.
> 3. I reload the page.
> 4. I don't see fields I added.
> 
> Screenshot 2024-10-20 at 16 06 58

We have not configured it like that. fields act like filters. those are not part of the configuration implementation.

**Comment by Matvey-Kuk** [comment]
@rajeshj11 wouldn't it be confusing for users?

**Comment by rajeshj11** [comment]
> @rajeshj11 wouldn't it be confusing for users?

Yes, I agree it could cause some confusion. We can definitely work on improving this.

**Comment by rajeshj11** [comment]
> @rajeshj11 wouldn't it be confusing for users?

@Matvey-Kuk The way we’ve implemented it—adding this in the config—would require significant effort. Should we create a subtask to address this?"

**Comment by rajeshj11** [comment]
> > @rajeshj11 wouldn't it be confusing for users?
> 
> @Matvey-Kuk The way we’ve implemented it—adding this in the config—would require significant effort. Should we create a subtask to address this?"

I have a quick solution for this. We can persist the widget table filters in local storage, so even after a refresh, it will retrieve the latest filters from local storage. what do you think?

**Comment by Matvey-Kuk** [comment]
@rajeshj11 the use-case I see is someone adding fields and sending the dashboard link to a colleague. It won't work in case it will be stored in the local storage.

**Comment by rajeshj11** [comment]
> @rajeshj11 the use-case I see is someone adding fields and sending the dashboard link to a colleague. It won't work in case it will be stored in the local storage.

Do you have any suggestions?

**Comment by vikashsprem** [comment]
Hey @Matvey-Kuk, we have changed according to your suggestion. Please review for more we have attached a demo video (`Some blinking issues in the video not in the actual application`) have a look.

https://www.loom.com/share/e8f8138b34da4836ae776088b11c406f

**Review by Matvey-Kuk — APPROVED** [review]
Great work, thank you!

**Review by talboren — APPROVED** [review]
LGTM

**Pull request merged**